File: leetcode/Permutations_2.cpp
class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int> &num) {
        if(num.empty()) return vector<vector<int>>();
        sort(num.begin(), num.end());
        
        vector<vector<int>> result;
        vector<int> sofar;
        permute_helper(result, sofar, num);
        
        return result;
    }
private:
    void permute_helper(vector<vector<int>> &result, vector<int> &sofar, vector<int> &rest){
        if(rest.empty()){
            result.push_back(sofar);
            return;
        }
        
        for(int i = 0; i < rest.size(); i++){
            // remove the duplication
            if(i > 0 && rest[i] == rest[i-1]) continue;
            int t = rest[i];
            
            sofar.push_back(t);
            auto it = rest.erase(rest.begin()+i);
            
            permute_helper(result, sofar, rest);
            
            sofar.pop_back();
            rest.insert(it, t);
        }
    }
};

File: leetcode/java/Permutations_2.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        if(nums == null) throw new IllegalArgumentException("x");
        if(nums.length == 0) return Collections.emptyList();
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        puh(result, new ArrayList<>(), Arrays.stream(nums).boxed().collect(Collectors.toList()));
        return result;
    }
    
    private void puh(List<List<Integer>> result, List<Integer> sofar, List<Integer> cands) {
        if(cands.isEmpty()) {
            result.add(new ArrayList<>(sofar));
            return;
        }
        
        for(int i = 0; i < cands.size(); i++){
            if(i > 0 && cands.get(i).equals(cands.get(i-1))) continue;
            Integer x = cands.get(i); 
            sofar.add(x);
            List<Integer> left = new ArrayList<>(cands);
            left.remove(x);
            puh(result, sofar, left);
            sofar.remove(sofar.size()-1);
        }
    }
}
